drgn/helpers/linux/llist.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2022, Oracle and/or its affiliates.
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+"""
+Lock-less NULL terminated single linked lists
+------------
+
+The ``drgn.helpers.linux.llist`` module provides helpers for working with the
+lockless single linked list implementations (``struct llist_head`` and ``struct
+llist_node``) in :linux:`include/linux/llist.h`.
+"""
+
+from typing import Iterator, Union
+
+from drgn import NULL, Object, Type, container_of
+
+
+__all__ = (
+    "llist_empty",
+    "llist_first_entry",
+    "llist_first_entry_or_null",
+    "llist_for_each",
+    "llist_for_each_entry",
+    "llist_is_singular",
+    "llist_next_entry",
+)
+
+def llist_empty(head: Object) -> bool:
+    """
+    Return whether an llist is empty.
+
+    :param head: ``struct llist_head *``
+    """
+    return not head.first
+
+
+def llist_is_singular(head: Object) -> bool:
+    """
+    Return whether an llist has only one element.
+
+    :param head: ``struct llist_head *``
+    """
+    return head.first.value_() != 0x0 and head.first.next.value_() == 0x0
+
+
+
+def llist_first_entry(head: Object, type: Union[str, Type], member: str) -> Object:
+    """
+    Return the first entry in an llist.
+
+    The list is assumed to be non-empty.
+
+    See also :func:`llist_first_entry_or_null()`.
+
+    :param head: ``struct llist_head *``
+    :param type: Entry type.
+    :param member: Name of llist_node member in entry type.
```

Comment:
Let's replace `Name of llist_node member` with `Name of ``struct llist_node`` member` everywhere in this file to be more explicit.

Reply:
Done